---
title: "R Notebook"
output: html_notebook
--- 
 
```{r}
g_p <- read.delim("/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/g_p.txt") %>% rename(gene = gene.x)

#variants <- read.delim('variant_quantificaton_0925.csv', sep = ",")
#variants <- read.delim('variant_quantificaton_GC.csv', sep = ",")

all_variants <- read.delim('variants_1004_batched.csv',  sep = ",") %>% filter(AC > 0) #actually have an allele

#tmsb4y_exons <- read.delim('TMSB4Y_variants_1003.csv',  sep = ",") %>% filter(AC > 0)
#all_variants <- read.delim('all_variants_GC.csv',  sep = ",")
#all_variants <- rbind(all_variants, tmsb4y_exons) 

summary(all_variants$AF) #mean is 0.2% 
all_var_tolook <- merge(all_variants, g_p, by = "gene")

```

```{r}
#all_variants <- unique(all_variants) #unique unless the variants have different vep annotations

unique(all_variants$IMPACT)
#all_variants$annotated_region <- ifelse((all_variants$region == "exon") & (all_variants$vep == "synonymous_variant"), "synonymous_exon", all_variants$region)

all_variants <- all_variants  %>% group_by(gene, region, IMPACT) %>% mutate(num_variants = length(position)) #%>% filter(AF > 0.00002)
#all_variants <- all_variants %>% filter(AF > 0.001) %>% group_by(gene, region) %>% mutate(num_variants = length(position))

ggplot(all_variants, aes(x=AF )) + 
  geom_histogram(bins = 400) +
  scale_y_log10() + 
  facet_wrap(~region, scales = 'free') +
  xlim(0,0.25)
  #xlim(0, 0.01)


summary_table_impact <- unique(all_variants %>% dplyr::select("gene", "region","IMPACT", "num_variants", "sum_len"))
summary_table_impact <- merge(summary_table_impact, g_p, by = "gene")

summary_table_impact$norm_nuc_pos <- summary_table_impact$num_variants / summary_table_impact$sum_len

all_variants <- all_variants %>% filter(AF > 0.00002) %>% group_by(gene, region) %>% mutate(num_variants = length(position))
summary_table_region <- unique(all_variants %>% dplyr::select("gene", "region", "num_variants", "sum_len"))
summary_table_region <- merge(summary_table_region, g_p, by = "gene")

summary_table_region$norm_nuc_pos <- summary_table_region$num_variants / summary_table_region$sum_len


#the 5' and 3' UTR's are likely impacting the actual AF for TMSB4Y - need to adjust for that --maybe there's a way to use the biotype information with a new seqlen calculation* 

```

```{r}

summary_table_impact %>% filter(region == "exon") %>% 
  ggplot(aes(x=type_of_gene, y = norm_nuc_pos)) + 
  facet_wrap(~IMPACT, scales = "free") + 
  geom_boxplot()+ 
  geom_point(aes(color = pair))

# variants
# g_p

ggplot(summary_table_region, aes(x=region, y = norm_nuc_pos)) +
  geom_boxplot() +
  geom_jitter() +
  facet_wrap(~gene, scales = 'free')

pdf('normalized_variants_1017.pdf', width = 6, height = 5)
ggplot(summary_table_region, aes(x=type_of_gene, y = norm_nuc_pos, color = pair)) +
  #geom_boxplot() +
  geom_jitter(width = 0.1) +
  facet_grid(~region) + 
  theme_pubclean()
dev.off()

ggplot(summary_table_region, aes(x=region, y = norm_nuc_pos, color = pair)) +
  #geom_boxplot() +
  #geom_jitter() +
  geom_point() +
  geom_line(aes(group = pair)) + 
  facet_wrap(~type_of_gene, scales = 'free')

summary_table1 <- unique(summary_table_region %>% select("gene", "region", "norm_nuc_pos")) 

# summary_table1 %<>% group_by(gene) %>% filter(n() == 3) %>% 
#   mutate(exon_ratio = norm_nuc_pos[which(region == "exon")]/norm_nuc_pos[which(region == "intron")], 
#          promoter_ratio = norm_nuc_pos[which(region == "promoter")]/norm_nuc_pos[which(region == "intron")])
summary_table1 %<>% group_by(gene) %>% filter(n() == 3) %>% 
  mutate(exon_ratio = norm_nuc_pos[which(region == "exon")]/norm_nuc_pos[which(region == "intron")], 
         promoter_ratio = norm_nuc_pos[which(region == "promoter")]/norm_nuc_pos[which(region == "intron")])
 summary_table1 <- merge(summary_table1, g_p, by = "gene")
summary_table1 <- unique(summary_table1 %>% dplyr::select("gene", "type_of_gene", "exon_ratio", "promoter_ratio"))

pdf('normalized_variants_ratios_1017.pdf', width = 6, height = 5)

summary_table1 %>% tidyr::pivot_longer(cols = c(3:4), names_to = "type", values_to = "ratio") %>% ggplot(aes(x = type_of_gene, y = ratio, color = type, label = gene)) + 
  #geom_boxplot() + 
  geom_point(position = position_dodge(width = 0.8)) + 
  ggrepel::geom_text_repel(position = position_dodge(width = 0.8), max.overlaps = 20) + 
  theme_pubclean() 
dev.off()
 
```
#looking at HIGH impact only
```{r}
all_variants <- read.delim('variants_1004_batched.csv',  sep = ",") %>% filter(AC > 0) %>% filter(qual == '[]') %>% filter(AF < 0.001) %>% filter(EXON != "")#actually have an allele


#MAKE THIS ALL VARIANTS IN EXONS
HIGH_impact_vars <- all_variants %>% filter(IMPACT == 'HIGH')
HIGH_impact_vars1 <- HIGH_impact_vars %>% group_by(SYMBOL) %>% mutate(added_AF = sum(AF))
write.csv(HIGH_impact_vars1, "/lab/solexa_page/hannah/1000genomes/HIGH_impact_vars.csv")

HIGH_impact_vars2 <- unique(HIGH_impact_vars1 %>% filter(region == 'exon') %>% select(SYMBOL, gene, sum_len, added_AF))
HIGH_impact_vars2$AFbylen_exon <- round((HIGH_impact_vars2$added_AF / HIGH_impact_vars2$sum_len), 10) #*10^7

write.csv(HIGH_impact_vars2, "/lab/solexa_page/hannah/1000genomes/HIGH_impact_vars_norm.csv")

MOD_impact_vars <- all_variants %>% filter(IMPACT == 'MODERATE')
MOD_impact_vars <- MOD_impact_vars %>% group_by(SYMBOL) %>% mutate(added_AF = sum(AF))
write.csv(MOD_impact_vars, "/lab/solexa_page/hannah/1000genomes/MOD_impact_vars.csv")

MOD_impact_vars <- unique(MOD_impact_vars %>% filter(region == 'exon') %>% select(SYMBOL, gene, sum_len, added_AF))
MOD_impact_vars$AFbylen_exon <- round((MOD_impact_vars$added_AF / MOD_impact_vars$sum_len), 10)

write.csv(MOD_impact_vars, "/lab/solexa_page/hannah/1000genomes/MOD_impact_vars_norm.csv")

###
all_variants <- read.delim('variants_1004_batched.csv',  sep = ",") %>% filter(AC > 0) %>% filter(qual == '[]') %>% filter(AF < 0.001) %>% filter(region == "promoter")#actually have an allele
unique(all_variants$IMPACT) #theyre all 'modifiers'
promoter_var <- all_variants %>% group_by(SYMBOL) %>% mutate(added_AF = sum(AF))
write.csv(promoter_var, "/lab/solexa_page/hannah/1000genomes/promoter_var.csv")
promoter_var <- unique(promoter_var %>% select(SYMBOL, gene, sum_len, added_AF))
promoter_var$AFbylen_promoter <- ((promoter_var$added_AF / promoter_var$sum_len )) #*10^7
write.csv(promoter_var, "/lab/solexa_page/hannah/1000genomes/promoter_var_norm.csv")



exon_variants <- read.delim('variants_1004_batched.csv',  sep = ",") %>% filter(AC > 0) %>% filter(qual == '[]') %>% filter(AF < 0.001) %>% filter(EXON != "") %>% filter(region == "exon")#actually have an allele 
exon_variants1 <- exon_variants %>% group_by(SYMBOL) %>% mutate(added_AF = sum(AF))
write.csv(exon_variants1, "/lab/solexa_page/hannah/1000genomes/exon_variants.csv")

exon_variants1 <- unique(exon_variants1 %>% filter(region == 'exon') %>% select(SYMBOL, gene, sum_len, added_AF))

exon_variants1$AFbylen_exon <- exon_variants1$added_AF / exon_variants1$sum_len #*10^7



m_afs <- full_join( exon_variants1, promoter_var, by = "gene")
#g_p$SYMBOL <- g_p$gene
m_afs <- merge(m_afs, g_p, by = "gene")
ggplot(m_afs, aes(x=AFbylen_promoter, y = AFbylen_exon)) + 
  geom_point() + 
  facet_wrap(~type_of_gene) + 
  stat_cor(method = "pearson") 


all_variants <- read.delim('variants_1004_batched.csv',  sep = ",") %>% filter(AC > 0) %>% filter(qual == '[]') %>% filter(AF < 0.001) %>% filter(region == "intron") #actually have an allele
#GC_introns <- read.delim('variants_1018_prom_norm_gc.csv', sep = ",") %>% filter(AC > 0) %>% filter(qual == '[]') %>% filter(AF < 0.001) %>% filter(region == "intron_GC")
#GC_intron_var <- GC_introns %>% group_by(gene) %>% mutate(added_AF = sum(AF))
#GC_intron_var <- unique(GC_intron_var %>% select(SYMBOL, gene, sum_len, added_AF))
#GC_intron_var$AFbylen_intron_GC <- ((GC_intron_var$added_AF / GC_intron_var$sum_len )) #*10^7

#unique(all_variants$IMPACT) #theyre all 'modifiers'
intron_var <- all_variants %>% group_by(gene) %>% mutate(added_AF = sum(AF))
write.csv(intron_var, "/lab/solexa_page/hannah/1000genomes/intron_var.csv")
intron_var <- unique(intron_var %>% select(SYMBOL, gene, sum_len, added_AF))
intron_var$AFbylen_intron <- ((intron_var$added_AF / intron_var$sum_len )) #*10^7
write.csv(intron_var, "/lab/solexa_page/hannah/1000genomes/intron_var_norm.csv")

m_afs <- merge(m_afs,intron_var, by = "gene")
#m_afs <- merge(m_afs, GC_intron_var, by = "gene")
m_afs$promoter_intron_ratio <- m_afs$AFbylen_promoter / m_afs$AFbylen_intron
m_afs$exon_intron_ratio <- m_afs$AFbylen_exon / m_afs$AFbylen_intron
#m_afs$promoter_intronGC_ratio <- m_afs$AFbylen_promoter / m_afs$AFbylen_intron_GC


m_afs1 <- unique(m_afs %>% select('gene', 'promoter_intron_ratio', 'exon_intron_ratio'))
m_afs2 <- unique(m_afs %>% select('gene', 'AFbylen_exon', 'AFbylen_intron', 'AFbylen_promoter'))

m_afs2_piv <- m_afs1 %>% tidyr::pivot_longer(c(2:3)) 
m_afs2_piv <-merge(m_afs2_piv, g_p, by = 'gene')

m_afs2_piv <- m_afs2_piv %>% filter(gene != 'NLGN4X') %>% filter(gene != 'NLGN4Y') %>% filter(gene != 'TMSB4X') %>% filter(gene != 'TMSB4Y') %>% filter(gene != 'TXLNG') %>% filter(gene != 'TXLNGY')

pdf("ratios_1017.pdf", width = 5, height = 5)
ggplot(m_afs2_piv, aes(x=type_of_gene, y=value, fill = name)) + 
  geom_violin() +
  geom_point(position = position_dodge(width = 0.9)) +
  theme_pubr() +
  ggrepel::geom_text_repel(aes(label=gene), position = position_dodge(width = 0.9))
dev.off()
#sum the allele frequencies? 
#sum the number of people with a LoF? 
# 50 100 = 0.5AF making the assumption that the same person does NOT have multiple LOF
# 50 100 = 0.5AF
#check that my gene names match

# LOF_impact_vars <- all_variants %>% filter(LoF == 'HC') #high conf LOF
# LOF_impact_vars %<>% group_by(SYMBOL) %>% mutate(added_AF = sum(AF))
# write.csv(LOF_impact_vars, "/lab/solexa_page/hannah/1000genomes/LoF_impact_vars.csv")


#can I normalize the sum of the frequencies by the number of possible bases assessed (exon length?) - 
#are my numbers for intron correct

#look at the promoters (number modifiers*)
#could imagine a situation where i take random 100bp sections 

```






<!-- #breakdown of what types of mutations -->

<!-- ```{r} -->

<!-- all_variants <- unique(all_variants) #unique unless the variants have different vep annotations -->

<!-- all_variants <- all_variants %>% filter(AF > 0.00002) %>% group_by(gene, region) %>% mutate(num_variants = length(position)) -->

<!-- all_var <- unique(all_variants %>% group_by(gene, vep) %>% mutate(sum = length(vep)) %>% select('vep', 'gene', 'sum')) -->

<!-- ``` -->


#gnomad mutational constraint 
```{r}
xy_pairs_gnomadV2 <- read.delim("/lab/solexa_page/hannah/1000genomes/xy_pairs_gnomadV2_constraint.txt")
merged_constr <- merge(xy_pairs_gnomadV2, g_p, by = "gene")

pdf('missense_constraint.pdf', width = 4, height = 5)
ggplot(merged_constr, aes(x=type_of_gene, y= miss_constraint)) + 
  geom_point() + 
  theme_pubr() + 
  ylab("missense Z score") + 
  geom_text_repel(aes(label = gene))
dev.off()


mtests <- merge(m_afs2, merged_constr, by = 'gene')
mtests_x <- mtests %>% filter(type_of_gene == "X_gene")
mtests_y <- mtests %>% filter(type_of_gene == "Y_gene")

pdf('pairs_Y.pdf', width = 10, height = 10)
ggpairs(mtests_y[,c(2:4, 8, 11, 14)], lower = list(continuous = "smooth", wrap = c(method = "loess")), cardinality_threshold = 20)
dev.off()
pdf('pairs_X.pdf', width = 10, height = 10)
ggpairs(mtests_x[,c(2:4, 8, 11, 14)], lower = list(continuous = "smooth", wrap = c(method = "loess")), cardinality_threshold = 20)
dev.off()

pdf('pairs_Y.pdf', width = 5, height = 5)
mtests_y %>% ggplot(aes(x=AFbylen_promoter, y=miss_constraint)) +
  geom_point() + 
  stat_cor(method = "pearson") +
  geom_text(aes(label=gene), position = position_dodge()) +
  theme_pubr()
dev.off()

```

































