---
title: "R Notebook"
output: html_notebook
---

```{r}
library(ggplot2)
library(ggpubr)
```
#gnomad V3
```{r}
g_p <- read.delim("/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/g_p.txt") %>% rename(gene = gene.x)
all_variants <- read.delim('variants_gnomad.csv',  sep = ",") #variants_gnomad.csv variants_0328_prom_norm_gc.csv
all_variants1 <- read.delim('variants_gnomad_CDS.csv',  sep = ",") 
all_variants <- rbind(all_variants, all_variants1)
all_variants <- all_variants %>% filter(AC > 0 & qual == "[]" & VARIANT_CLASS == "SNV") #allele count greater than 1 & VARIANT_CLASS == "SNV" & AF < 0.5 & AF > 0.05 & AF < 0.95 

all_var_tolook <- merge(all_variants, g_p, by = "gene")
all_var_tolook$X <- NULL
all_var_tolook <- all_var_tolook %>% select(c("gene", "chrom", "position", "region", "sum_len", "pair", "type_of_gene", "direction")) %>% distinct()

all_var_1 <- all_var_tolook %>% group_by(gene, region) %>% mutate(sum_v = n()) 

all_var_1$ratio <- all_var_1$sum_v / all_var_1$sum_len

all_var_1 <- all_var_1 %>% select("gene", "region", "ratio", "sum_v") %>% distinct() %>% filter(region != "intron_GC") #ned something there 
write.table(x = all_var_1, file = '/lab/solexa_page/hannah/1000genomes/polymorphism_gnomad.txt', quote = FALSE, sep = "\t",  row.names = FALSE)

```

#gnomad V4
```{r}

g_p <- read.delim("/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/g_p.txt") %>% rename(gene = gene.x)
all_variants <- read.delim('variants_gnomad_V4.csv',  sep = ",") #variants_gnomad.csv variants_0328_prom_norm_gc.csv
all_variants <- all_variants %>% filter(AC > 0 & qual == "[]") # & VARIANT_CLASS == "SNV") #allele count greater than 1 & VARIANT_CLASS == "SNV" & AF < 0.5 & AF > 0.05 & AF < 0.95
nrow(all_variants)

# all_variants <- all_variants %>% filter(AC_XY != '[0]') # & qual == "[]" & VARIANT_CLASS == "SNV") #allele count greater than 1 & VARIANT_CLASS == "SNV" & AF < 0.5 & AF > 0.05 & AF < 0.95
# nrow(all_variants)

all_var_tolook <- merge(all_variants, g_p, by = "gene")
all_var_tolook$X <- NULL
all_var_tolook <- all_var_tolook %>% select(c("gene", "chrom", "position", "region", "sum_len", "pair", "type_of_gene", "direction")) %>% distinct()

all_var_1 <- all_var_tolook %>% group_by(gene, region) %>% mutate(sum_v = n()) 

all_var_1$ratio <- all_var_1$sum_v / all_var_1$sum_len

all_var_1 <- all_var_1 %>% select("gene", "region", "ratio", "sum_v") %>% distinct() %>% filter(region != "intron_GC") #ned something there 
write.table(x = all_var_1, file = '/lab/solexa_page/hannah/1000genomes/polymorphism_gnomad_V4.txt', quote = FALSE, sep = "\t",  row.names = FALSE)



```


#1000 genomes 
```{r}
g_p <- read.delim("/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/g_p.txt") %>% rename(gene = gene.x)
all_variants <- read.delim('variants_0328_prom_norm_gc.csv',  sep = ",") #variants_gnomad.csv variants_0328_prom_norm_gc.csv


all_var_tolook <- merge(all_variants, g_p, by = "gene")
all_var_tolook$X <- NULL
all_var_tolook <- all_var_tolook %>% select(c("gene", "chrom", "position", "region", "sum_len", "pair", "type_of_gene", "direction")) %>% distinct()

all_var_1 <- all_var_tolook %>% group_by(gene, region) %>% mutate(sum_v = n()) 

all_var_1$ratio <- all_var_1$sum_v / all_var_1$sum_len

all_var_1 <- all_var_1 %>% select("gene", "region", "ratio", "sum_v") %>% distinct() %>% filter(region != "intron_GC") #ned something there 
write.table(x = all_var_1, file = '/lab/solexa_page/hannah/1000genomes/polymorphism_1000genomes.txt', quote = FALSE, sep = "\t",  row.names = FALSE)



```



```{r}
# all_var_2 <-all_var_1 %>% filter(gene != "ZFX" & gene != "ZFY")
all_var_2 <- all_var_1 %>% group_by(gene) %>% filter(n() == 3) %>% #n() == 4 |
  mutate(exon_ratio = ratio[which(region == "CDS")]/ratio[which(region == "intron")], 
         promoter_ratio = ratio[which(region == "promoter")]/ratio[which(region == "intron")])

summary_table1 <- merge(all_var_2, g_p, by = "gene")
summary_table1 <- unique(summary_table1 %>% dplyr::select("gene", "type_of_gene", "exon_ratio", "promoter_ratio"))
 
```



```{r}

summary_table1 %>% tidyr::pivot_longer(cols = c(3:4), names_to = "type", values_to = "ratio") %>% ggplot(aes(x = type_of_gene, y = ratio, color = type, label = gene)) + 
  #geom_boxplot() + 
  geom_point(position = position_dodge(width = 0.8)) + 
  ggrepel::geom_text_repel(position = position_dodge(width = 0.8), max.overlaps = 20) + 
  theme_pubr() 

```









