{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c55aa666-d81d-48a9-bf4a-f4d453ce466a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import Bio\n",
    "from Bio import SeqIO \n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "import glob\n",
    "from Bio import SeqRecord\n",
    "import pandas as pd\n",
    "####pip install genomicranges #this worked \n",
    "import genomicranges as gr\n",
    "from Bio.Seq import Seq\n",
    "####from Bio.SeqRecord import SeqRecord \n",
    "import pyranges as pr\n",
    "from Bio import AlignIO, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fac0a00a-b8d6-4771-baef-8eaeb534475b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def make_sbatch_file(filename):\n",
    "    \n",
    "    my_list = [\"#!/bin/bash\", \n",
    "               \"#SBATCH --job-name=all_bz\", \n",
    "               \"#SBATCH --nodes=1\",  \n",
    "               \"#SBATCH --ntasks=1\",                     \n",
    "               \"#SBATCH --cpus-per-task=10\",              \n",
    "               \"#SBATCH --mem=20gb\",                    \n",
    "               \"#SBATCH --partition=20\",                \n",
    "               \"##SBATCH --output all_bz-%j.out\",  \n",
    "               \"#SBATCH --mail-type=ALL\",               \n",
    "               \"#SBATCH --mail-user=hlharris@wi.mit.edu\"] \n",
    "    \n",
    "    with open(filename, \"w\") as file: \n",
    "        for item in my_list:\n",
    "            file.write(item + '\\n') \n",
    "    \n",
    "def calc_zeros(alignment): \n",
    "    count = 0 \n",
    "    return [0 if base == \"-\" else (count := count + 1) for base in alignment[0]]\n",
    "\n",
    "\n",
    "def crop_alignment(start, end, alignment):\n",
    "        \n",
    "    zeros_seq = calc_zeros(alignment)\n",
    "    #print(zeros_seq) #MIGHT NEED TO SUBTRACT 1 FROM EACH \n",
    "    try:\n",
    "        ix_start = zeros_seq.index(start)\n",
    "        #print(ix_start)\n",
    "        ix_end = zeros_seq.index(end) #pretty elegant\n",
    "        #print(ix_end)\n",
    "        cropped_alignment = alignment[:, ix_start:ix_end]\n",
    "        return cropped_alignment\n",
    "    except: \n",
    "        return None\n",
    "    \n",
    "    \n",
    "def concat_alignment(gtf_file, alignment):\n",
    "   # print(alignment)\n",
    "    complete_align_type = MultipleSeqAlignment([]) \n",
    "    #add the groups to complete align: \n",
    "    for recordix in range(len(alignment)): \n",
    "        #new_record = SeqRecord.SeqRecord(\"\") \n",
    "        new_record = SeqRecord(\"\")\n",
    "        new_record.id = alignment[recordix].id \n",
    "        new_record.seq = Seq(\"\") #added this \n",
    "\n",
    "        complete_align_type.append(new_record)\n",
    "    \n",
    "   # print(complete_align_type)\n",
    "    for index, row in gtf_file.iterrows():     \n",
    "        crop_align = crop_alignment(row[3], row[4], alignment)\n",
    "        #print(crop_align)\n",
    "        if crop_align is not None: \n",
    "            for recordix in range(len(crop_align)): \n",
    "                complete_align_type[recordix].seq += crop_align[recordix].seq #append additional sequence\n",
    "        \n",
    "    return complete_align_type\n",
    "\n",
    "def has_ACGT(sequence):\n",
    "    count_ACGT = sum(c in 'ACTG' for c in sequence)\n",
    "    return count_ACGT > len(sequence) / 4 #true or false"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "753e3d88-daf3-4d5c-a297-c77411159db9",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DDX3X\n",
      "Submitted batch job 792806\n",
      "DDX3X\n",
      "EIF1AX\n",
      "Submitted batch job 792807\n",
      "EIF1AX\n",
      "KDM5C\n",
      "Submitted batch job 792808\n",
      "KDM5C\n",
      "KDM6A\n",
      "Submitted batch job 792809\n",
      "KDM6A\n",
      "RPS4X\n",
      "Submitted batch job 792810\n",
      "RPS4X\n",
      "USP9X\n",
      "Submitted batch job 792811\n",
      "USP9X\n",
      "ZFX\n",
      "Submitted batch job 792812\n",
      "ZFX\n"
     ]
    }
   ],
   "source": [
    "#generate alignment \n",
    "\n",
    "genes = ['DDX3X', 'EIF1AX', 'KDM5C', 'KDM6A', 'RPS4X', 'USP9X', 'ZFX']\n",
    "\n",
    "\n",
    "for gene in genes: \n",
    "\n",
    "    tree = '(humanmasked (bull dog))'\n",
    "    \n",
    "    os.chdir('/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/'+ gene)\n",
    "\n",
    "    subprocess.run([\"touch\", \"all_bz.log\"])  #making a new file \n",
    "    \n",
    "    with open(\"all_bz.log\", \"w\") as log_file:\n",
    "        result = subprocess.run(['all_bz', '-', tree], stdout=log_file, stderr=subprocess.PIPE, check=True)\n",
    "        \n",
    "    make_sbatch_file(\"testj1.sh\") #make sbatch file - it makes a new one each time through so you dont have to delete the old one \n",
    "    print(gene) \n",
    "    with open(\"testj1.sh\", \"a\") as file: #append to the file \n",
    "        file.write(\"bash \") \n",
    "        file.write(\"all_bz.log\" + \"\\n\")\n",
    "        file.write(\"tba '\" + tree + \"' *.*.maf tba.maf >&tba.log\" + \"\\n\") \n",
    "        file.write(\"maf_project tba.maf humanmasked '\" + tree + \"' > human_proj.maf\" + \"\\n\") #project from the perspective of the X chr \n",
    "        file.write(\"msa_view -o FASTA human_proj.maf > \" + gene + \"_msa.fa\")\n",
    "        \n",
    "    subprocess.run([\"sbatch\", 'testj1.sh'], stderr=subprocess.PIPE) \n",
    "    print(gene)\n",
    "    os.chdir('/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/')  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6f52832f-18e9-4585-bba7-3a0856c6cbac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get alignments with no editing of gaps w trimal\n",
    "genes = ['DDX3X', 'EIF1AX', 'KDM5C', 'KDM6A', 'RPS4X', 'USP9X', 'ZFX']\n",
    "regions = [\"promoter\", \"exon\", \"intron\"]   \n",
    "\n",
    "for gene in genes: \n",
    "    alignment = AlignIO.read(\"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + gene + \"_msa.fa\", \"fasta\")\n",
    "    \n",
    "    gtf_file = pd.read_csv(\"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/\" + gene + \"_gtf_all103023.txt\", delimiter=\"\\t\", header = None) \n",
    "    \n",
    "    for region in regions: \n",
    "        \n",
    "        gtf_file1 = gtf_file[gtf_file[2] == region] \n",
    "        #print(gtf_file1)\n",
    "        returned_align = concat_alignment(gtf_file1,alignment) \n",
    "        #print(gene_pair, returned_align, region)\n",
    "        SeqIO.write(returned_align, \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + gene + \"_\" + region + \"_msa.fa\", \"fasta\") #as fasta\n",
    "        SeqIO.write(returned_align, \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + gene + \"_\" + region + \"_msa.phy\", \"phylip\") #as phylip\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2d2826a-e6a5-4d10-a691-4b8ef6d117a2",
   "metadata": {},
   "source": [
    "## generate percent calculations for regions/alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "049824c8-84e5-42f4-bcf9-3a76bb90c5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def edit_alignment(records_X, gene, region, specy):\n",
    "\n",
    "    new_alignment = MultipleSeqAlignment([])\n",
    "    \n",
    "    if specy == \"bull\":\n",
    "        records_x_lost = [0,1] #take edit the alignment \n",
    "    else:\n",
    "        records_x_lost = [0,2]\n",
    "        \n",
    "    for ix in records_x_lost:\n",
    "        \n",
    "        new_record = records_X[ix]\n",
    "     \n",
    "        new_alignment.append(new_record)\n",
    "    \n",
    "    #print(new_alignment)\n",
    "    SeqIO.write(new_alignment, \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + \"humanmasked.\" + specy + region + \"_gaps_msa.phy\", \"phylip\") #as phylip\n",
    "    subprocess.run(['/lab/page_scratch/hannah/trimal/source/trimal', '-in', \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + \"humanmasked.\" + specy + region + \"gaps_msa.phy\", '-out',   \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + \"humanmasked.\" + specy + region + \"_gaps_msa.phy\", '-noallgaps'],  stderr=subprocess.PIPE)\n",
    "\n",
    "    records_X = list(SeqIO.parse( \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + \"humanmasked.\" + specy + region + \"_gaps_msa.phy\", \"phylip\"))\n",
    "    \n",
    "    \n",
    "    new_alignment = MultipleSeqAlignment([])\n",
    "    records_x_lost = [0,1] #take edit the alignment since now its only 2 sequences long\n",
    "    \n",
    "    for ix in records_x_lost:\n",
    "    # Replace '*' with '-' in the sequence\n",
    "        \n",
    "        modified_seq  = str(records_X[ix].seq).replace('-', 'K')\n",
    "        modified_seq = str(modified_seq).replace('*', '-')\n",
    "        modified_seq = str(modified_seq).replace('K', '*') #replacing the - w stars - want to keep the gaps for the calculations of % alignments \n",
    "        new_record = records_X[ix]\n",
    "        new_record.seq = Seq(modified_seq)     \n",
    "        new_alignment.append(new_record)\n",
    "                \n",
    "    print(gene, region, new_alignment, \"IN LOOP\")  \n",
    "    \n",
    "    SeqIO.write(new_alignment, \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + specy + region + \"replgaps_msa.phy\", \"phylip\") #as phylip\n",
    "\n",
    "    subprocess.run(['/lab/page_scratch/hannah/trimal/source/trimal', '-in', \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + specy + region + \"replgaps_msa.phy\", '-out', \"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + specy + region + \"nogaps_msa.phy\", '-gt', '1'],  stderr=subprocess.PIPE)\n",
    "        \n",
    "    try:\n",
    "        new_alignment = list(SeqIO.parse(\"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + specy + region + \"nogaps_msa.phy\", \"phylip\"))    \n",
    "        return new_alignment\n",
    "          \n",
    "    except: \n",
    "        return []\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "acd504c8-11f0-4733-9d43-3f1fc5b25937",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DDX3X exon SingleLetterAlphabet() alignment with 2 rows and 2634 columns\n",
      "GGATGAGTCATGTGGCAGTGGAAAATGCGCTCGGGCTGGACCAG...CAT humanmaske\n",
      "GGATGAGTCATGTGGCGGTGGAAAATGCGCTCGGGCTGGACCAG...AAT dog IN LOOP\n",
      "DDX3X exon SingleLetterAlphabet() alignment with 2 rows and 2634 columns\n",
      "GGATGAGTCATGTGGCAGTGGAAAATGCGCTCGGGCTGGACCAG...CAT humanmaske\n",
      "GGATGAGTCATGTGGCGGTGGAAAATGCGCTCGGGCTGGACCAG...--- bull IN LOOP\n",
      "DDX3X intron SingleLetterAlphabet() alignment with 2 rows and 22393 columns\n",
      "TTTTTTCTCCCAA***CTACTTGTGAGTAGTTCCTGGTTATAAA...AAT humanmaske\n",
      "TTTTTTCCCCCTGGTATTGCTTGCAG*TAGTTTCTGGTT***AG...*** dog IN LOOP\n",
      "DDX3X intron SingleLetterAlphabet() alignment with 2 rows and 22393 columns\n",
      "TTTTTTCTCCCAA***CTACTTGTGAGTAGTTCCTGGTTATAAA...AAT humanmaske\n",
      "AGTTCCCTCCTCG***TTGCTTGTAAATAGTTT**GTTT***AG...--- bull IN LOOP\n",
      "DDX3X promoter SingleLetterAlphabet() alignment with 2 rows and 514 columns\n",
      "ATCTCCAAGCCC*GGGAACTGAGAGAGGAA*GAAGAGTGAAGGC...GCT humanmaske\n",
      "GCCCCCAAGCCC*GAGGACTCAGAGAAAAAAGAA**********...GCT dog IN LOOP\n",
      "DDX3X promoter SingleLetterAlphabet() alignment with 2 rows and 514 columns\n",
      "ATCTCCAAGCCC*GGGAACTGAGAGAGGAA*GAAGAGTGAAGGC...GCT humanmaske\n",
      "CTTTCAAAGTCTAGGGGACTGAGAGAGAAG*GAA**********...ATT bull IN LOOP\n",
      "EIF1AX exon SingleLetterAlphabet() alignment with 2 rows and 432 columns\n",
      "CATGCCCAAGAATAAAGGTAAAGGAGGTAAAAACAGACGCAGGG...GAT humanmaske\n",
      "CATGCCCAAGAATAAAGGTAAAGGAGGTAAAAATAGACGCAGAG...GAT dog IN LOOP\n",
      "EIF1AX exon SingleLetterAlphabet() alignment with 2 rows and 432 columns\n",
      "CATGCCCAAGAATAAAGGTAAAGGAGGTAAAAACAGACGCAGGG...GAT humanmaske\n",
      "CATGCCTAAGAATAAAGGTAAAGGAGGTAAAAACAGACGCAGAG...GAT bull IN LOOP\n",
      "EIF1AX intron SingleLetterAlphabet() alignment with 2 rows and 13480 columns\n",
      "CCTTCCCGCCCTGCCTGCCCGCGGCTGGGCGCCCCACACGCGAC...TTC humanmaske\n",
      "CCTCTCCGTAGTGCCTGTGCGCGGCTGCACCGTACGGACGCGGT...*** dog IN LOOP\n",
      "EIF1AX intron SingleLetterAlphabet() alignment with 2 rows and 13480 columns\n",
      "CCTTCCCGCCCTGCCTGCCCGCGGCTGGGCGCCCCACACGCGAC...TTC humanmaske\n",
      "CCTCTCCGGCCTGCCTTCACACAGCTAGACTATCCGGACGCGAT...TTC bull IN LOOP\n",
      "EIF1AX promoter SingleLetterAlphabet() alignment with 2 rows and 542 columns\n",
      "CTACTCCGGAGGCCGAGGCAGGAGAATCGCTTGAACCCGGGCGG...GCC humanmaske\n",
      "--------------------------------------------...GCC dog IN LOOP\n",
      "EIF1AX promoter SingleLetterAlphabet() alignment with 2 rows and 542 columns\n",
      "CTACTCCGGAGGCCGAGGCAGGAGAATCGCTTGAACCCGGGCGG...GCC humanmaske\n",
      "--------------------------------------------...GCC bull IN LOOP\n",
      "KDM5C exon SingleLetterAlphabet() alignment with 2 rows and 4966 columns\n",
      "CATGGAGCCGGGGTCCGACGATTTCCTACCGCCACCGGAGTGCC...CTG humanmaske\n",
      "CATGGAGCCGGGATCCGACGACTTCCTGCCGCCGCCGGAGTGCC...CTG dog IN LOOP\n",
      "KDM5C exon SingleLetterAlphabet() alignment with 2 rows and 4966 columns\n",
      "CATGGAGCCGGGGTCCGACGATTTCCTACCGCCACCGGAGTGCC...CTG humanmaske\n",
      "CATGGAGCCGGGGTCTGACGACTTTCTACCGCCGCCGGAGTGCC...CTG bull IN LOOP\n",
      "KDM5C intron SingleLetterAlphabet() alignment with 2 rows and 24973 columns\n",
      "AGGGGGTGGCTGGGACCGTTGTGGTTGCGGCG*CGGATCGGAAA...CTC humanmaske\n",
      "ACGGGATTGCTGGGACTGCTGTGGATGTGGCGGTGGGTCGGAGA...CTC dog IN LOOP\n",
      "KDM5C intron SingleLetterAlphabet() alignment with 2 rows and 24973 columns\n",
      "AGGGGGTGGCTGGGACCGTTGTGGTTGCGGCG*CGGATCGGAAA...CTC humanmaske\n",
      "ATGGGGTGGCTAGCACCGTTGTGGATGCGGCG*CGGGTCGGAGA...CTT bull IN LOOP\n",
      "KDM5C promoter SingleLetterAlphabet() alignment with 2 rows and 510 columns\n",
      "AGTACATCTCCCCCCATCCCGCGGGCCAGCATAAGGGGTCCTTT...GCT humanmaske\n",
      "CCTATGTCTTTCCCAA*CCTGTTTGTCAGAATAAGGGATCCCTC...GCT dog IN LOOP\n",
      "KDM5C promoter SingleLetterAlphabet() alignment with 2 rows and 510 columns\n",
      "AGTACATCTCCCCCCATCCCGCGGGCCAGCATAAGGGGTCCTTT...GCT humanmaske\n",
      "TGTATGTCTCTCGTAA*CCTGCGGGCCTGCAGAAGGG*TCCCTT...GGT bull IN LOOP\n",
      "KDM6A exon SingleLetterAlphabet() alignment with 2 rows and 4499 columns\n",
      "CCATGAAATCCTGCGGAGTGTNNNNNNNNNNNNNNNNNNNNNNN...CAT humanmaske\n",
      "CCATGAAATCCTGCGGAGTGT***********************...CAT dog IN LOOP\n",
      "KDM6A exon SingleLetterAlphabet() alignment with 2 rows and 4499 columns\n",
      "CCATGAAATCCTGCGGAGTGTNNNNNNNNNNNNNNNNNNNNNNN...CAT humanmaske\n",
      "CCATGAAATCCAGCGGAGTGT***********************...CAT bull IN LOOP\n",
      "KDM6A intron SingleLetterAlphabet() alignment with 2 rows and 257496 columns\n",
      "GCA*CCTCGGTTGACGATGGTCGAGGGGCTTCCGG*AAACTATT...T*T humanmaske\n",
      "GCACCCCGGGCTAACGATGGCCGAGGGGCTTCCGGAAAACTATT...TCT dog IN LOOP\n",
      "KDM6A intron SingleLetterAlphabet() alignment with 2 rows and 257496 columns\n",
      "GCA*CCTCGGTTGACGATGGTCGAGGGGCTTCCGG*AAACTATT...T*T humanmaske\n",
      "GCA*CCTGGGCTAACGAGGGCCGAGGGGCTTCCGGAAAACTATT...TCT bull IN LOOP\n",
      "KDM6A promoter SingleLetterAlphabet() alignment with 2 rows and 548 columns\n",
      "CCCCTGGAGCGCGCGCGGGACGGCCGGAGCCGCGGTCTGGGCAC...CAA humanmaske\n",
      "CCCCCGGA****GCGCGGGACTGCCGGGGCAGTGGTCTGGGCCC...CAA dog IN LOOP\n",
      "KDM6A promoter SingleLetterAlphabet() alignment with 2 rows and 548 columns\n",
      "CCCCTGGAGCGCGCGCGGGACGGCCGGAGCCGCGGTCTGGGCAC...CAA humanmaske\n",
      "CACGCGGA****GCACAGGA*TGCCGGAGCCGCGGTCGGGGACC...CAA bull IN LOOP\n",
      "RPS4X exon SingleLetterAlphabet() alignment with 2 rows and 789 columns\n",
      "CATGGCTCGTGGTCCCAAGAAGCATCTGAAGCGGGTGGCAGCTC...TGG humanmaske\n",
      "CATGGCTCGTGGTCCCAAGAAGCATCTGAAGCGTGTAGCAGCTC...TGG dog IN LOOP\n",
      "RPS4X exon SingleLetterAlphabet() alignment with 2 rows and 789 columns\n",
      "CATGGCTCGTGGTCCCAAGAAGCATCTGAAGCGGGTGGCAGCTC...TGG humanmaske\n",
      "CATGGCTCGGGGTCCCAAGAAGCACCTGAAACGCGTAGCAGCTC...TGG bull IN LOOP\n",
      "RPS4X intron SingleLetterAlphabet() alignment with 2 rows and 3319 columns\n",
      "GCAGAGGCGAAGGGATCACGGCCTCCGTGGTAGGCCGGGACGGG...AAA humanmaske\n",
      "GTGGAGGCAAAGGGACTGCAGCCTGCGAGTTGG*CCCAGG****...*** dog IN LOOP\n",
      "RPS4X intron SingleLetterAlphabet() alignment with 2 rows and 3319 columns\n",
      "GCAGAGGCGAAGGGATCACGGCCTCCGTGGTAGGCCGGGACGGG...AAA humanmaske\n",
      "GCCGAGGCGAAGGGAGAGGGGCTAGTGGGGTGGGCACAGGCTGG...AAA bull IN LOOP\n",
      "RPS4X promoter SingleLetterAlphabet() alignment with 2 rows and 517 columns\n",
      "CTGCCTCAGCTTCCCGAGTAGCTGGGACTACAGGCGCCCGCCAC...CCC humanmaske\n",
      "--------------------------------------------...CCT dog IN LOOP\n",
      "RPS4X promoter SingleLetterAlphabet() alignment with 2 rows and 517 columns\n",
      "CTGCCTCAGCTTCCCGAGTAGCTGGGACTACAGGCGCCCGCCAC...CCC humanmaske\n",
      "*TGATTCAGATATATACATATAGACATATATAGATATATATTA*...CCT bull IN LOOP\n",
      "USP9X exon SingleLetterAlphabet() alignment with 2 rows and 7712 columns\n",
      "GTATGACAGCCACGACTCGTGGCTCTCCGGTCGGAGGGAATGAC...ATC humanmaske\n",
      "GTATGACAGCCACGACTCGTGGCTCTCCAGTAGGAGGGAATGAC...--- dog IN LOOP\n",
      "USP9X exon SingleLetterAlphabet() alignment with 2 rows and 7712 columns\n",
      "GTATGACAGCCACGACTCGTGGCTCTCCGGTCGGAGGGAATGAC...ATC humanmaske\n",
      "GTATGACAGCCACGACTCGTGGCTCTCCAGTAGGAGGGAATGAC...ATC bull IN LOOP\n",
      "USP9X intron SingleLetterAlphabet() alignment with 2 rows and 104991 columns\n",
      "ACTTGAGGTCAAGAGTTTGAGACCAGCCTGGTCAACAGGGTGGA...GTC humanmaske\n",
      "--------------------------------------------...--- dog IN LOOP\n",
      "USP9X intron SingleLetterAlphabet() alignment with 2 rows and 104991 columns\n",
      "ACTTGAGGTCAAGAGTTTGAGACCAGCCTGGTCAACAGGGTGGA...GTC humanmaske\n",
      "--------------------------------------------...GTC bull IN LOOP\n",
      "USP9X promoter SingleLetterAlphabet() alignment with 2 rows and 689 columns\n",
      "GGTGT**GCCAGATCTTTTCGGGCCTCCCTTCCCCAACTT****...TTT humanmaske\n",
      "GGCGTTTGCCAGATCCTTCCGGACTTCCCTTCCCCCACTTTTCC...TTT dog IN LOOP\n",
      "USP9X promoter SingleLetterAlphabet() alignment with 2 rows and 689 columns\n",
      "GGTGT**GCCAGATCTTTTCGGGCCTCCCTTCCCCAACTT****...TTT humanmaske\n",
      "GGCAT**GCCAGATCCTTTCGGGCCGCCCCGCCCCCGCTTCTGC...TTT bull IN LOOP\n",
      "ZFX exon SingleLetterAlphabet() alignment with 2 rows and 2596 columns\n",
      "CCATGGATGAAGATGGGCTTGAATTACA***ACAAGAGCCAAAC...TGC humanmaske\n",
      "CCATGGATGAAGATGGACTTGAATTACAACCACAAGAGCCAAAC...TGC dog IN LOOP\n",
      "ZFX exon SingleLetterAlphabet() alignment with 2 rows and 2596 columns\n",
      "CCATGGATGAAGATGGGCTTGAATTACA***ACAAGAGCCAAAC...TGC humanmaske\n",
      "CCATGGATGAAGATGGACTTGAATTACAACCACAAGAACCAAAC...TGC bull IN LOOP\n",
      "ZFX intron SingleLetterAlphabet() alignment with 2 rows and 38728 columns\n",
      "ATCTTGTTATCCTGTCAAAATTCACAGGTGTTACGGTGAGCACA...AGG humanmaske\n",
      "ATCTT*TTATCCTGTCAAGCCT*ACATTTGTTAAAGGGGGCATA...AGG dog IN LOOP\n",
      "ZFX intron SingleLetterAlphabet() alignment with 2 rows and 38728 columns\n",
      "ATCTTGTTATCCTGTCAAAATTCACAGGTGTTACGGTGAGCACA...AGG humanmaske\n",
      "ATCTTATTATCCTGTCAAACCT*ACAGATGTTGAAGTGAGTGCA...AGG bull IN LOOP\n",
      "ZFX promoter SingleLetterAlphabet() alignment with 2 rows and 592 columns\n",
      "CCTCGGCCAGGACCGGCGC*GAGAGGCGGTAGGTCTCCAGGGAG...GGC humanmaske\n",
      "CCTCGGCCAAGACAGGCGCGGAGAGGCGGTAGGTCTCCGGGGAG...GGC dog IN LOOP\n",
      "ZFX promoter SingleLetterAlphabet() alignment with 2 rows and 592 columns\n",
      "CCTCGGCCAGGACCGGCGC*GAGAGGCGGTAGGTCTCCAGGGAG...GGC humanmaske\n",
      "CCTGGGCCAGGACCGGCGCGGAGAGGCGGTAGGTCTCCGGGGAG...GGC bull IN LOOP\n"
     ]
    }
   ],
   "source": [
    "regions = [\"exon\", \"intron\", \"promoter\"]  \n",
    "genes = ['DDX3X', 'EIF1AX', 'KDM5C', 'KDM6A', 'RPS4X', 'USP9X', 'ZFX']\n",
    "species = [\"dog\", \"bull\"]\n",
    "\n",
    "p = []\n",
    "r = []\n",
    "perc = []\n",
    "speci = []\n",
    "\n",
    "for gene in genes: \n",
    "   \n",
    "    \n",
    "    for region in regions: \n",
    "           # print(records_X, \"here\")     \n",
    "            records_X = list(SeqIO.parse('/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/' + gene + \"/\" + gene + \"_\" + region + \"_msa.fa\", \"fasta\")) \n",
    "            #print(records_X[0])\n",
    "            for specy in species: \n",
    "                new_alignment = edit_alignment(records_X, gene, region, specy)\n",
    "            \n",
    "                if new_alignment:\n",
    "                    x_gene = new_alignment[0] \n",
    "                    y_gene = new_alignment[1] \n",
    "                    num_matches = sum(a == b for a, b in zip(x_gene, y_gene))\n",
    "             \n",
    "                    percent_identity = (num_matches / len(x_gene)) * 100\n",
    "                    p.append(gene)\n",
    "                    r.append(region) \n",
    "                    speci.append(specy)\n",
    "                    perc.append(percent_identity)\n",
    "                else: \n",
    "                    print(pair, region, \"NONE\")\n",
    "                \n",
    "\n",
    "\n",
    "df = pd.DataFrame((list(zip(p,r,perc, speci))), columns = ['pair', 'region', 'percent', \"species\"]) \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "39972764-69b7-4867-80a6-de1e3193acd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df\n",
    "df.to_csv('/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/percent_alignment_1211_dogbull.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f0197872-c610-4392-83e1-254ad76797c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = []\n",
    "gc = []\n",
    "speci = []\n",
    "regions1 = []\n",
    "\n",
    "regions = [\"exon\", \"intron\", \"promoter\"]  \n",
    "genes = ['DDX3X', 'EIF1AX', 'KDM5C', 'KDM6A', 'RPS4X', 'USP9X', 'ZFX']\n",
    "species = [\"dog\", \"bull\"]\n",
    "\n",
    "#regions = [\"promoter\"]  \n",
    "#genes = ['DDX3X']\n",
    "#species = [\"dog\"]\n",
    "\n",
    "\n",
    "for gene in genes: \n",
    "    for region in regions:\n",
    "        for specy in species: \n",
    "            records = AlignIO.read(\"/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/\" + gene + \"/\" + \"humanmasked.\" + specy + region + \"_gaps_msa.phy\", \"phylip\")\n",
    "            human_seq = str(records[0].seq)\n",
    "            animal_seq = str(records[1].seq)\n",
    "            \n",
    "            h_gc = ((human_seq.count('G') + human_seq.count('C')) / (human_seq.count('G') + human_seq.count('C') + human_seq.count('A') + human_seq.count('T'))) *100\n",
    "            a_gc = ((animal_seq.count('G') + animal_seq.count('C')) / (animal_seq.count('G') + animal_seq.count('C') + animal_seq.count('A') + animal_seq.count('T'))) *100\n",
    "            \n",
    "            gc.append(h_gc)\n",
    "            g.append(gene)\n",
    "            speci.append('human')\n",
    "            \n",
    "            regions1.append(region)\n",
    "            \n",
    "            gc.append(a_gc)\n",
    "            g.append(gene)\n",
    "            speci.append(specy)\n",
    "            regions1.append(region) \n",
    "\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "541fcb49-558e-4ecf-a7fa-f35ebea8524b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame((list(zip(g,gc,speci, regions1))), columns = ['gene', 'gc', 'species', 'region']) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "1adb29f9-3f22-4180-ae73-84d00b105035",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gc</th>\n",
       "      <th>species</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX3X</td>\n",
       "      <td>46.253544</td>\n",
       "      <td>human</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DDX3X</td>\n",
       "      <td>44.920235</td>\n",
       "      <td>dog</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DDX3X</td>\n",
       "      <td>46.253544</td>\n",
       "      <td>human</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DDX3X</td>\n",
       "      <td>44.752658</td>\n",
       "      <td>bull</td>\n",
       "      <td>exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DDX3X</td>\n",
       "      <td>43.763954</td>\n",
       "      <td>human</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ZFX</td>\n",
       "      <td>37.405241</td>\n",
       "      <td>bull</td>\n",
       "      <td>intron</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ZFX</td>\n",
       "      <td>77.701149</td>\n",
       "      <td>human</td>\n",
       "      <td>promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>ZFX</td>\n",
       "      <td>77.333333</td>\n",
       "      <td>dog</td>\n",
       "      <td>promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>ZFX</td>\n",
       "      <td>77.701149</td>\n",
       "      <td>human</td>\n",
       "      <td>promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>ZFX</td>\n",
       "      <td>76.862745</td>\n",
       "      <td>bull</td>\n",
       "      <td>promoter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>84 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene         gc species    region\n",
       "0   DDX3X  46.253544   human      exon\n",
       "1   DDX3X  44.920235     dog      exon\n",
       "2   DDX3X  46.253544   human      exon\n",
       "3   DDX3X  44.752658    bull      exon\n",
       "4   DDX3X  43.763954   human    intron\n",
       "..    ...        ...     ...       ...\n",
       "79    ZFX  37.405241    bull    intron\n",
       "80    ZFX  77.701149   human  promoter\n",
       "81    ZFX  77.333333     dog  promoter\n",
       "82    ZFX  77.701149   human  promoter\n",
       "83    ZFX  76.862745    bull  promoter\n",
       "\n",
       "[84 rows x 4 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "49adbfea-deaf-48ce-ba6b-80d1ea80ef58",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('/lab/solexa_page/hannah/220516_mpra/msa/long_alignments/multiz_dog_bull/percent_GC_1211_dogbull.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a96c1c83-dbc6-4022-a6d3-d7bdf213da62",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd9d1af8-9119-44d7-a1f2-2b75e13f134b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "233fa373-6c97-4a43-8421-bb2ab05cdca7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ab2ef12-33b5-4372-bf87-7fcca7e47f29",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f43651e5-d6e6-4c97-b872-eb7b1538853c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
